src/lib/falcor/falcorUtils.js

```diff
@@ -40,95 +40,110 @@ export function validateFalcorPathSets(falcorPathSets) {
+    }
+
+    return undefined;
+  }, object);
+}
+
+export function pathSetsInCache(cache, falcorPathSets) {
+  function handleCheckingSingleKey(curObject, nextRemainingKeySets, key) {
+    if (!curObject.hasOwnProperty(key)) {
+      return false;
+    }
+    else {
+      const val = curObject[key];
+      if (val.$type) {
+        switch (val.$type) {
+          case "error":
+          case "atom":
+            return nextRemainingKeySets.length === 0;
+          case "ref":
+            return checkSinglePathSetInCache(followPath(val.value, cache), nextRemainingKeySets);
+          default:
+            throw new Error("pathSetsInCache encountered unexpected type. Type found was: " + val.$type);
         }
-        end = key.to;
-      }
-      else if (key.hasOwnProperty("length")) {
-        end = start+key.length-1;
       }
       else {
-        throw new Error("Falcor Range must have either 'to' or 'length' properties at falcor KeySet: " + JSON.stringify(key));
+        return checkSinglePathSetInCache(val, nextRemainingKeySets);
       }
-      for (let i = start; i <= end; i++) {
-        if (curObject.hasOwnProperty(i)) {
-          if (!recurseOnCache(curObject[i], nextRemainingKeySets)) {
-            return false;
+    }
+  }
+
+  function checkSinglePathSetInCache(curObject, remainingKeySets) {
+    if (remainingKeySets.length === 0) {
+      return true;
+    }
+    const nextRemainingKeySets = remainingKeySets.slice(1);
+    let nextKeySet = remainingKeySets[0];
+    // This is to avoid code duplication for when
+    // it is just a single key instead of an
+    // array of keys. We don't want to handle that
+    // case seperately so just make an array with one value
+    if (!(nextKeySet instanceof Array)) {
+      nextKeySet = [nextKeySet];
+    }
+    return nextKeySet.every((keyOrRange) => {
+      if (keyOrRange !== null && typeof keyOrRange === "object") {
+        // keyOrRange is a range
+        let start = 0;
+        if (keyOrRange.hasOwnProperty("from")) {
+          start = keyOrRange.from;
+        }
+        let end;
+        if (keyOrRange.hasOwnProperty("to")) {
+          if (keyOrRange.hasOwnProperty("length")) {
+            throw new Error("Falcor Range cannot have both 'to' and 'length' properties at falcor KeySet: " + JSON.stringify(keyOrRange));
           }
+          end = keyOrRange.to;
+        }
+        else if (keyOrRange.hasOwnProperty("length")) {
+          end = start+keyOrRange.length-1;
         }
         else {
-          return false;
+          throw new Error("Falcor Range must have either 'to' or 'length' properties at falcor KeySet: " + JSON.stringify(keyOrRange));
         }
-      }
-    }
-    else {
-      if (curObject.hasOwnProperty(key)) {
-        if (!recurseOnCache(curObject[key], nextRemainingKeySets)) {
-          return false;
+        for (let i = start; i <= end; i++) {
+          if (!handleCheckingSingleKey(curObject, nextRemainingKeySets, i)) {
```

Comment:
One thing we need to think about here, is the "overtfetching" case. Where we fetch like 10 authors for one article, even though there are only 2, just to be safe. In this case, it is still loaded... This is a tricky edge case to handle. Do you have ideas  how we can validate that these are cached already in a correct manner?